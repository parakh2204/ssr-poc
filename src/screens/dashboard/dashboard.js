/**
 * The main dashboard/landing screen, which shows the default listing of the champions, with an autocomplete search bar
 * that allows the user to search the record with suggestions and tapping on any champion from the list displays it's details.
 * The sorting function is also implemented, that is totally offline, and sorts the complete list of champions alphabettically in either order.
 *
 * Tapping on any of the list item, displays the details of the  champion, with the option to add that champion in the favourite list or remove
 * from the same.
 *
 * The count of champions in the favourite list is also visible on the the top panel, list item. Tapping on which, takes the user to the Watch list page.
 */

import React, { useState, useEffect, useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Avatar, Spin, Input, Card, Badge, Menu, Dropdown } from "antd";
import { useHistory } from "react-router-dom";
import useWindowDimensions from "./../../commonComponents/customhooks/useWindowDimensions";
import {
  Container,
  LoaderWrap,
  StyledAutoComplete,
  CardWrap,
  StyledList,
  AutoCompleteList,
  HeaderWrap,
  CardImage,
} from "./dashboardStyles";
import DetailModal from "../../commonComponents/detailModal/detailModal";
import {
  getChampionsList,
  searchChampions,
} from "./../../redux/actions/listingAction";
import {
  addToWatchList,
  removeFromWatchList,
} from "../../redux/actions/watchListActions";

import {
  AppstoreAddOutlined,
  DeleteOutlined,
  ProfileOutlined,
  SortAscendingOutlined,
} from "@ant-design/icons";
export const Dashboard = (props) => {
  const reducers = useSelector((state) => state);
  const history = useHistory();
  const { listingReducer, watchListReducer } = reducers;
  const { watchList } = watchListReducer;
  const { championsList, loading, searchedResults } = listingReducer;
  const [query, setQuery] = useState("");
  const [modal2Visible, setModal2Visible] = useState(false);
  const [champDetail, setChampDetail] = useState(null);
  const [sortLoader, setSortLoader] = useState(false);

  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getChampionsList());
  }, []);
  const { Meta } = Card;
  const { height } = useWindowDimensions();
  /**
   * maps the results returned from the search api in the autocomplete box with image and name of the champion
   */
  const searchResult = useCallback(() => {
    return searchedResults.map((_, idx) => {
      const category = _.name;
      return {
        value: category,
        label: (
          <AutoCompleteList>
            <span>{category}</span>
            <Avatar size={"large"} alt={_.image_url} src={_.image_url} />
          </AutoCompleteList>
        ),
        completeRecord: _,
      };
    });
  }, [searchedResults]);
  /**
   * sets the data according to the query made in the autocomplete bar
   */
  useEffect(() => {
    setOptions(query ? searchResult() : []);
  }, [query, searchResult]);

  const [options, setOptions] = useState([]);

  /**
   *
   * @param {string} value - the query string made by the user
   */
  const handleSearch = (value) => {
    // eslint-disable-next-line no-unused-expressions
    value && value !== "" ? dispatch(searchChampions(value)) : null;
    setQuery(value);
  };

  const menu = (
    <Menu>
      <Menu.ItemGroup title="Sort Items">
        <Menu.Item>
          <div
            onClick={() => {
              championsList.sort((a, b) =>
                a.name > b.name ? 1 : b.name > a.name ? -1 : 0
              );
              setSortLoader(true);
              setTimeout(() => {
                setSortLoader(false);
              }, 400);
            }}
          >
            A to Z
          </div>
        </Menu.Item>
        <Menu.Item>
          <div
            onClick={() => {
              championsList.sort((a, b) =>
                a.name > b.name ? -1 : b.name > a.name ? 1 : 0
              );
              setSortLoader(true);
              setTimeout(() => {
                setSortLoader(false);
              }, 400);
            }}
          >
            Z to A
          </div>
        </Menu.Item>

        <Menu.Item>
          <div>Close</div>
        </Menu.Item>
      </Menu.ItemGroup>
    </Menu>
  );
  return (
    <Container height={height}>
      <HeaderWrap>
        <StyledAutoComplete
          data-testid="searchBar"
          dropdownMatchSelectWidth={252}
          options={options}
          onSearch={handleSearch}
          onSelect={(item, option) => {
            setChampDetail(option.completeRecord);
            setModal2Visible(true);
          }}
        >
          <Input.Search
            size="large"
            placeholder="Search Champions"
            enterButton
          />
        </StyledAutoComplete>
        <Badge
          onClick={() => history.push("/watchlist")}
          count={watchList.length}
        >
          <ProfileOutlined data-testid="listIcon" style={{ fontSize: 30 }} />
        </Badge>
        <Dropdown overlay={menu} placement="bottomLeft">
          <SortAscendingOutlined
            data-testid="sortIcon"
            style={{ fontSize: 30 }}
          />
        </Dropdown>
      </HeaderWrap>
      {loading || sortLoader ? (
        <LoaderWrap>
          <Spin size="large" tip="Loading Champions" />
        </LoaderWrap>
      ) : null}

      {championsList && (
        <StyledList
          grid={{
            gutter: 0,
            column: 4,
            xs: 1,
            sm: 1,
            md: 2,
            lg: 3,
            xl: 4,
            xxl: 4,
          }}
          size="small"
          pagination={{
            pageSize: 8,
            showSizeChanger: false,
          }}
          dataSource={championsList}
          align="middle"
          renderItem={(item) => (
            <CardWrap>
              <Card
                align="left"
                hoverable
                bordered
                cover={
                  <CardImage
                    height={height * 0.5}
                    onClick={() => {
                      setChampDetail(item);
                      setModal2Visible(true);
                    }}
                    alt="big_image_url"
                    src={item.big_image_url}
                  />
                }
                actions={[
                  <AppstoreAddOutlined
                    onClick={() => dispatch(addToWatchList(item))}
                    key="addToWatchList"
                  />,

                  <DeleteOutlined
                    onClick={() => dispatch(removeFromWatchList(item))}
                    key="removeFromWatchList"
                  />,
                ]}
                style={{ width: "100%" }}
              >
                <Meta
                  onClick={() => {
                    setChampDetail(item);
                    setModal2Visible(true);
                  }}
                  avatar={<Avatar src={item.image_url} />}
                  title={item.name}
                  description={"Attack Damage: " + item.attackdamage}
                />
              </Card>
            </CardWrap>
          )}
        />
      )}

      {champDetail && (
        <DetailModal
          champDetail={champDetail}
          okFunction={() => {
            dispatch(addToWatchList(champDetail));
            setModal2Visible(false);
          }}
          cancelFunction={() => {
            dispatch(removeFromWatchList(champDetail, true));
            setModal2Visible(false);
          }}
          okText={"Add to Fav."}
          cancelText={"Remove from Fav."}
          visibleStatus={modal2Visible}
        />
      )}
    </Container>
  );
};
export default Dashboard;
