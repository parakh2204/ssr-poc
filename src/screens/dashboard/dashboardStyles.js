import styled from "styled-components";
import { AutoComplete, List } from "antd";
import { device } from "./../../utils/helpers";

export const Container = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  background-color: white;
  overflow-y: scroll;
  height: ${(props) => props.height + "px"};
  margin: ${(props) => props.margin};
`;

export const LoaderWrap = styled.div`
  display: flex;
  flex: 1;

  align-items: center;
  justify-content: center;
`;

export const CardWrap = styled.div`
  display: flex;
  margin: 2%;

  border-radius: 1%;

  background-color: transparent;
  width: 68%;
  overflow-x: hidden;
  box-shadow: 0px 0px 10px 2px grey;
  &:hover {
    box-shadow: 0px 0px 15px 5px grey;
  }

  @media ${device.mobileS} {
    width: 90%;
  }
  @media ${device.tablet} {
    width: 68%;
  }
  @media ${device.desktop} {
    width: 68%;
  }
`;

export const StyledAutoComplete = styled(AutoComplete)`
  width: 70%;
  align-self: center;
  @media ${device.mobileS} {
    width: 70%;
  }
  @media ${device.tablet} {
    width: 70%;
  }
  @media ${device.desktop} {
    width: 70%;
  }
`;

export const StyledList = styled(List)`
  display: flex;
  flex-direction: column;
  align-items: center;
  position: relative;
`;

export const ModalWrap = styled.div`
  display: flex;
  flex: 1;
  flex-direction: row;
  justify-content: space-between;
  margin-bottom: 4%;
`;

export const BoldName = styled.b`
  font-size: 18px;
`;

export const AutoCompleteList = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const HeaderWrap = styled.div`
  display: flex;
  justify-content: space-evenly;
  background-color: white;
  flex-direction: row;
  align-items: center;
  position: sticky;
  top: 1px;
  z-index: 100;
  padding-top: 6%;
  padding-bottom: 1%;

  @media ${device.mobileS} {
    padding-top: 6%;
  }
  @media ${device.tablet} {
    padding-top: 6%;
  }
  @media ${device.desktop} {
    padding-top: 1%;
  }
`;

export const CardImage = styled.img`
  height: ${(props) => props.height + "px"};
`;
