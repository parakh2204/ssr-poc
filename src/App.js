import "./App.css";
import React from "react";
import { createStore, applyMiddleware, compose } from "redux";
import { Provider } from "react-redux";
import combinedReducer from "./../src/redux/combinedReducers";

import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { Dashboard } from "../src/screens/dashboard/dashboard";
import { WatchList } from "../src/screens/watchList/watchList";
const ReduxThunk = require("redux-thunk").default;

let composeEnhancer = compose;
export const store = createStore(
  //creates an instance of store to be used globally
  combinedReducer, // pass in the combined reducers
  composeEnhancer(applyMiddleware(ReduxThunk)) // use the middleware for async actions
);
export default function App(props) {
  return (
    <Provider store={store}>
      <Router>
        <Switch>
          <Route path="/" exact component={Dashboard} />
          <Route path="/watchlist" component={WatchList} />
          <Route path="/article/:id" component={Dashboard} />
        </Switch>
      </Router>
    </Provider>
  );
}