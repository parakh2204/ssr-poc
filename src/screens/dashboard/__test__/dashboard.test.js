import { render } from "@testing-library/react";
import * as React from "react";
import { Provider } from "react-redux";
import { store } from "./../../../App";
import { Dashboard } from "../dashboard";
import { shallow } from "./../../../utils/enzyme";
import { fireEvent, screen } from "@testing-library/react";

import renderer from "react-test-renderer";
import { CardImage } from "../dashboardStyles";

describe("Dashboard Component", () => {
  let DashComponent = (
    <Provider store={store}>
      <Dashboard />
    </Provider>
  );

  it("renders search bar correctly", () => {
    const { getByTestId } = render(DashComponent);
    expect(getByTestId("searchBar")).toBeVisible();
  });

  it("renders listing button correctly", () => {
    const { getByTestId } = render(DashComponent);
    expect(getByTestId("listIcon")).toBeVisible();
  });

  it("renders sort icon correctly", () => {
    const { getByTestId } = render(DashComponent);
    expect(getByTestId("sortIcon")).toBeVisible();
  });

  it("list is not available initially", () => {
    const wrapper = shallow(DashComponent);
    expect(wrapper.prop("championsList")).toBeUndefined();
  });

  test("search bar  must set search query correctly", () => {
    render(DashComponent);
    const searchBar = screen.getByPlaceholderText("Search Champions");
    expect(searchBar).toHaveValue(""); // checks for initial value to be ""

    fireEvent.change(searchBar, { target: { value: "testquery" } });
    expect(searchBar).toHaveValue("testquery"); // checks by passing a value testquery
  });

  it("listed item's click event works fine", () => {
    const mockCallBack = jest.fn();

    const listItem = shallow(<CardImage onClick={mockCallBack}>Ok!</CardImage>);
    listItem.find("img").simulate("click");
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it("matches snapshot perfectly", () => {
    const tree = renderer.create(DashComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
