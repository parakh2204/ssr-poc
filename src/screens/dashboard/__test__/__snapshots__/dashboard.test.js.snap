// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component matches snapshot perfectly 1`] = `
<div
  className="sc-bdnxRM vtpak"
  height={768}
>
  <div
    className="sc-gKAaRy fkerGe"
  >
    <div
      className="ant-select ant-select-auto-complete sc-hKFxyN jMSORU ant-select-single ant-select-customize-input ant-select-show-search"
      data-testid="searchBar"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
    >
      <div
        className="ant-select-selector"
        onClick={[Function]}
        onMouseDown={[Function]}
      >
        <span
          className="ant-select-selection-search"
        >
          <span
            className="ant-input-group-wrapper ant-input-group-wrapper-lg ant-input-search ant-input-search-large ant-input-search-with-button ant-select-selection-search-input"
            style={
              Object {
                "opacity": null,
              }
            }
          >
            <span
              className="ant-input-wrapper ant-input-group"
            >
              <input
                aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                aria-autocomplete="list"
                aria-controls="rc_select_TEST_OR_SSR_list"
                aria-haspopup="listbox"
                aria-owns="rc_select_TEST_OR_SSR_list"
                autoComplete="off"
                className="ant-input ant-input-lg"
                id="rc_select_TEST_OR_SSR"
                onBlur={[Function]}
                onChange={[Function]}
                onCompositionEnd={[Function]}
                onCompositionStart={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onPaste={[Function]}
                placeholder="Search Champions"
                readOnly={false}
                role="combobox"
                style={null}
                type="search"
                unselectable={null}
                value=""
              />
              <span
                className="ant-input-group-addon"
              >
                <button
                  className="ant-btn ant-btn-primary ant-btn-lg ant-input-search-button"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  type="button"
                >
                  <span
                    aria-label="search"
                    className="anticon anticon-search"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="search"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                      />
                    </svg>
                  </span>
                </button>
              </span>
            </span>
          </span>
        </span>
        <span
          className="ant-select-selection-placeholder"
        />
      </div>
    </div>
    <span
      className="ant-badge"
      onClick={[Function]}
    >
      <span
        aria-label="profile"
        className="anticon anticon-profile"
        data-testid="listIcon"
        role="img"
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        <svg
          aria-hidden="true"
          data-icon="profile"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656zM492 400h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H492c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8zm0 144h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H492c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8zm0 144h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H492c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8zM340 368a40 40 0 1080 0 40 40 0 10-80 0zm0 144a40 40 0 1080 0 40 40 0 10-80 0zm0 144a40 40 0 1080 0 40 40 0 10-80 0z"
          />
        </svg>
      </span>
    </span>
    <span
      aria-label="sort-ascending"
      className="anticon anticon-sort-ascending ant-dropdown-trigger"
      data-testid="sortIcon"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="img"
      style={
        Object {
          "fontSize": 30,
        }
      }
    >
      <svg
        aria-hidden="true"
        data-icon="sort-ascending"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M839.6 433.8L749 150.5a9.24 9.24 0 00-8.9-6.5h-77.4c-4.1 0-7.6 2.6-8.9 6.5l-91.3 283.3c-.3.9-.5 1.9-.5 2.9 0 5.1 4.2 9.3 9.3 9.3h56.4c4.2 0 7.8-2.8 9-6.8l17.5-61.6h89l17.3 61.5c1.1 4 4.8 6.8 9 6.8h61.2c1 0 1.9-.1 2.8-.4 2.4-.8 4.3-2.4 5.5-4.6 1.1-2.2 1.3-4.7.6-7.1zM663.3 325.5l32.8-116.9h6.3l32.1 116.9h-71.2zm143.5 492.9H677.2v-.4l132.6-188.9c1.1-1.6 1.7-3.4 1.7-5.4v-36.4c0-5.1-4.2-9.3-9.3-9.3h-204c-5.1 0-9.3 4.2-9.3 9.3v43c0 5.1 4.2 9.3 9.3 9.3h122.6v.4L587.7 828.9a9.35 9.35 0 00-1.7 5.4v36.4c0 5.1 4.2 9.3 9.3 9.3h211.4c5.1 0 9.3-4.2 9.3-9.3v-43a9.2 9.2 0 00-9.2-9.3zM416 702h-76V172c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v530h-76c-6.7 0-10.5 7.8-6.3 13l112 141.9a8 8 0 0012.6 0l112-141.9c4.1-5.2.4-13-6.3-13z"
        />
      </svg>
    </span>
  </div>
  <div
    className="sc-gtsrHT gWQSSh"
  >
    <div
      className="ant-spin ant-spin-lg ant-spin-spinning ant-spin-show-text"
    >
      <span
        className="ant-spin-dot ant-spin-dot-spin"
      >
        <i
          className="ant-spin-dot-item"
        />
        <i
          className="ant-spin-dot-item"
        />
        <i
          className="ant-spin-dot-item"
        />
        <i
          className="ant-spin-dot-item"
        />
      </span>
      <div
        className="ant-spin-text"
      >
        Loading Champions
      </div>
    </div>
  </div>
</div>
`;
